tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,262 @@ def option_price(*,
+        [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0], # up and in call
+        [0, 0, 1, 1, -1, -1, 1, 1, 0, 0, 1, 1], # down and in put
+        [0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0], # down and in call
+        [0, 0, 1, 1, 0, 0, -1, -1, 0, 0, 1, 1], # up and out put
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1], # up and out call
+        [1, 1, -1, -1, 1, 1, -1, -1, 0, 0, 1, 1], # down and out put
+        [1, 1, 0, 0, -1, -1, 0, 0, 0, 0, 1, 1]]) # down and out call
+
+    mask_matrix_lower_strike = tf.constant([
+        [0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0], # up and in put
+        [0, 0, 1, 1, -1, -1, 1, 1, 1, 1, 0, 0], # up and in call
+        [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0], # down and in put
+        [1, 1, -1, -1, 0, 0, 1, 1, 1, 1, 0, 0], # down and in call
+        [1, 1, 0, 0, -1, -1, 0, 0, 0, 0, 1, 1], # up and out put
+        [1, 1, -1, -1, 1, 1, -1, -1, 0, 0, 1, 1], # up and out call
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1], # down and out put
+        [0, 0, 1, 1, 0, 0, -1, -1, 0, 0, 1, 1]]) # down and out call
+
+    # Get mask for each option
+    masks_lower = tf.gather(mask_matrix_lower_strike, indices)
+    masks_greater = tf.gather(mask_matrix_greater_strike, indices)
+    strikes_greater = tf.expand_dims(strikes > barriers, axis=-1)
+    masks = tf.where(strikes_greater, masks_greater, masks_lower)
+    masks = tf.cast(masks, dtype=dtype)
+
+    call_or_put = tf.where(tf.equal(is_call_options, 0), -1., 1.)
+    below_or_above = tf.where(tf.equal(is_knock_out, 0), -1., 1.)
+
+    # Calculate params for integrals
+    time_volatilities = volatilities*expiries**.5
+    mu = (discount_rates-continuous_dividends)-((volatilities**2)/2)
+    lamda = 1+(mu/(volatilities**2))
+    x = tf.math.log(spots/strikes)/(
+        time_volatilities)+(lamda*time_volatilities)
+    x1 = tf.math.log(spots/barriers)/(
+        time_volatilities)+(lamda*time_volatilities)
+    y = tf.math.log((barriers**2)/(spots*strikes))/(
+        time_volatilities)+(lamda*time_volatilities)
+    y1 = tf.math.log(barriers/spots)/(
+        time_volatilities)+(lamda*time_volatilities)
+    b = ((mu**2)+(2*(volatilities**2)*discount_rates))/(volatilities**2)
+    z = tf.math.log(barriers/spots)/(
+        time_volatilities)+(b*time_volatilities)
+    a = mu/(volatilities**2)
+
+
+    # Other params used for integrals
+    discount_rates_exponent = tf.math.exp(
+        -1*discount_rates*expiries,
+        name="discount_rates_exponent")
+    continuous_dividends_exponent = tf.math.exp(
+        -1*continuous_dividends*expiries,
+        name="continuous_dividends_exponent")
+    barriers_ratio = tf.math.divide(barriers, spots,
+                                    name="barriers_ratio")
+    spots_term = call_or_put*spots*continuous_dividends_exponent
+    strikes_term = call_or_put*strikes*discount_rates_exponent
+    # Constructing Matrix with first and second algebraic terms for each integral
+    terms_mat = tf.stack(
```

Comment:
Can you please add a comment on what `terms_mat ` shape is, as well as `cdf_mat`?

Reply:
Done!